physics/cooling/include/cooling/cooling.hpp

```diff
@@ -26,219 +26,109 @@
-    RT_HeI_ionization_rate   = RT_HeI_ionization_rate_gr;
-    RT_HeII_ionization_rate  = RT_HeII_ionization_rate_gr;
-    RT_H2_dissociation_rate  = RT_H2_dissociation_rate_gr;
-    H2_self_shielding_length = H2_self_shielding_length_gr;
-}
-
-struct grackle_options
+    struct GlobalValues;
+    struct chemistry_data_;
+
+template <typename T>
+struct Cooler
 {
-    std::optional<int>   with_radiative_cooling                 = std::nullopt;
-    std::optional<int>   primordial_chemistry                   = std::nullopt;
-    std::optional<int>   h2_on_dust                             = std::nullopt;
-    std::optional<int>   metal_cooling                          = std::nullopt;
-    std::optional<int>   cmb_temperature_floor                  = std::nullopt;
-    std::optional<int>   UVbackground                           = std::nullopt;
-    std::optional<float> UVbackground_redshift_on               = std::nullopt;
-    std::optional<float> UVbackground_redshift_fullon           = std::nullopt;
-    std::optional<float> UVbackground_redshift_drop             = std::nullopt;
-    std::optional<float> UVbackground_redshift_off              = std::nullopt;
-    std::optional<float> Gamma                                  = std::nullopt;
-    std::optional<int>   three_body_rate                        = std::nullopt;
-    std::optional<int>   cie_cooling                            = std::nullopt;
-    std::optional<int>   h2_optical_depth_approximation         = std::nullopt;
-    std::optional<int>   photoelectric_heating_rate             = std::nullopt;
-    std::optional<int>   Compton_xray_heating                   = std::nullopt;
-    std::optional<int>   LWbackground_intensity                 = std::nullopt;
-    std::optional<int>   LWbackground_sawtooth_suppression      = std::nullopt;
-    std::optional<int>   use_volumetric_heating_rate            = std::nullopt;
-    std::optional<int>   use_specific_heating_rate              = std::nullopt;
-    std::optional<int>   use_radiative_transfer                 = std::nullopt;
-    std::optional<int>   radiative_transfer_coupled_rate_solver = std::nullopt;
-    std::optional<int>   radiative_transfer_intermediate_step   = std::nullopt;
-    std::optional<int>   radiative_transfer_hydrogen_only       = std::nullopt;
-    std::optional<int>   H2_self_shielding                      = std::nullopt;
-    std::optional<int>   dust_chemistry                         = std::nullopt;
+    Cooler();
+    ~Cooler();
+
+    void init(const double ms_sim, const double kp_sim, const int comoving_coordinates,
+              const std::optional<chemistry_data_> grackleOptions     = std::nullopt,
+              const std::optional<std::string>    grackleOptionsFile = std::nullopt,
+              const std::optional<double>         t_sim              = std::nullopt);
+
+    GlobalValues& get_global_values();
+    chemistry_data_ getDefaultChemistryData();
```

Comment:
You can't provide the values as an opaque type, and, perhaps you shouldn't.
`cool_particle` needs to be a member function of `Cooler`, such that `GlobalValues` only needs to be in the `Impl` class on the other side of the compiler firewall.